crates/flux-infer/src/refine_tree.rs

```diff
@@ -780,3 +791,328 @@ impl RefineCtxtTrace {
+        self.kv_lhs
+            .get(kvid)
+            .unwrap_or(&FxHashSet::default())
+            .iter()
+            .copied()
+            .collect()
+    }
+
+    fn build(&mut self, node: &Node, ctx: &mut Vec<KVid>) {
+        let n = ctx.len();
+        match &node.kind {
+            NodeKind::Head(expr, _) => {
+                let mut vertices = vec![];
+                expr_vertices(expr, &mut vertices);
+                for rhs in &vertices {
+                    self.insert_edge(ctx, *rhs);
+                }
+            }
+            NodeKind::Assumption(expr) => {
+                let mut vertices = vec![];
+                expr_vertices(expr, &mut vertices);
+                for v in &vertices {
+                    if let VertexId::KVar(kvid) = v {
+                        ctx.push(*kvid);
+                    }
+                }
+            }
+            _ => {}
+        };
+
+        for child in &node.children {
+            self.build(&child.borrow(), ctx);
+        }
+
+        ctx.truncate(n); // restore ctx
+    }
+
+    fn insert_edge(&mut self, lhs: &[KVid], rhs: VertexId) {
+        let edge_id = CstrId(self.edges.len());
+
+        // Create and insert edge
+        let edge_info = CstrInfo { lhs: lhs.iter().copied().collect(), rhs };
+        self.edges.insert(edge_id, edge_info);
+
+        // Add edge_id to kv_lhs for each kvar in lhs
+        for kvid in lhs {
+            self.kv_lhs.entry(*kvid).or_default().insert(edge_id);
+        }
+        // Add edge_id to kv_rhs for rhs kvar
+        if let VertexId::KVar(kvid) = rhs {
+            self.kv_rhs.entry(kvid).or_default().insert(edge_id);
+        }
+    }
+
+    fn kvars(&self) -> impl Iterator<Item = KVid> + '_ {
+        self.kv_lhs
+            .keys()
+            .chain(self.kv_rhs.keys())
+            .copied()
+            .collect::<FxHashSet<_>>()
+            .into_iter()
+    }
+}
+
+/// helper function to split an `Expr` into its constituent `VertexId`s
+fn expr_vertices(expr: &Expr, vertices: &mut Vec<VertexId>) {
+    match expr.kind() {
+        ExprKind::BinaryOp(BinOp::And, e1, e2) => {
+            expr_vertices(e1, vertices);
+            expr_vertices(e2, vertices);
+        }
+        ExprKind::KVar(kvar) => vertices.push(VertexId::KVar(kvar.kvid)),
+        _ => vertices.push(VertexId::Conc),
+    }
+}
+/// Wrapper type for edge IDs, uniquely generated for each "Head"
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+struct CstrId(usize);
+
+#[derive(Debug)]
+struct CstrInfo {
+    /// assumptions
+    lhs: FxHashSet<KVid>,
+    /// head
+    rhs: VertexId,
+}
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum VertexId {
+    /// KVar
+    KVar(KVid),
+    /// Concrete
+    Conc,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Label {
+    /// Kvar can be solved to false
+    Bot,
+    /// Kvar can be solved to true
+    Top,
+}
+
+struct Assignment {
+    inner: FxHashMap<KVid, Label>,
```

Comment:
We always assign everything to the same level, so we can make this a set and store the `Label` as a separate field.